pkg/controller/controller.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2017 by the contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	authenticatorV1alpha1 "github.com/kubernetes-sigs/aws-iam-authenticator/pkg/apis/authenticator.aws/v1alpha1"
+	authenticatorclient "github.com/kubernetes-sigs/aws-iam-authenticator/pkg/client/clientset/versioned"
+	authenticatorinformerV1alpha1 "github.com/kubernetes-sigs/aws-iam-authenticator/pkg/client/informers/externalversions/authenticator.aws/v1alpha1"
+	"github.com/kubernetes-sigs/aws-iam-authenticator/pkg/config"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/workqueue"
+
+	"fmt"
+	"github.com/sirupsen/logrus"
+	"time"
+)
+
+// New creates a new controller to listen to CRDs
+func New(c *config.Config) (*Controller, error) {
+	_, clientset, authenticatorset, err := createClient(c.Kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	iamidentityInformer := authenticatorinformerV1alpha1.NewIAMIdentityInformer(authenticatorset, "kube-system", 10*time.Minute, cache.Indexers{})
+	ctrl := &Controller{
+		Config:              c,
+		Clientset:           clientset,
+		Authenticatorset:    authenticatorset,
+		IAMIdentityInformer: iamidentityInformer,
+		Queue:               workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "authenticatorsync"),
+	}
+
+	logrus.Print("setting up iamidentity listeners")
+	iamidentityInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				identity := obj.(*authenticatorV1alpha1.IAMIdentity).DeepCopy()
+				logrus.Infof("iamidentity '%s' added", identity.Name)
+				roleMap := config.RoleMapping{
+					RoleARN:  identity.Spec.RoleARN,
+					Username: identity.Spec.Username,
+					Groups:   identity.Spec.Groups,
+				}
+				c.AddRoleMapping(roleMap)
+				logrus.Infof("Roles with one added: %+v", c.CanonicalRoleMappings)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				oldIdentity := oldObj.(*authenticatorV1alpha1.IAMIdentity).DeepCopy()
```

Comment:
Do you have a pointer on why the DeepCopy() is needed?  We are not modifying it and just copying it into another object?